_posts/2016-03-12-make-your-data-obey.markdown

```diff
@@ -0,0 +1,198 @@
+    // Password now specifies the modifier 'encrypt' should be used to
+    // encrypt the password
+    password: { type: 'string', modifier: 'encrypt', require: true }
+})
+{% endhighlight %}
+
+Now not only does our validation method create a UUID, now it also encrypts the password for us.
+
+But what about that `email` property? Probably want to make sure it's unique...
+
+## Types
+
+Having that `email` property contain one that's already in the datasource would be bad news, right? Some databases do provide uniqueness integrity checking, but there are a lot that don't. Luckily, `types` are extensible with Obey, so we can add our own check:
+
+{% highlight javascript %}
+import obey from 'obey'
+import uuid from 'node-uuid'
+import argon2 from 'argon2'
+// Here's our db reference again...
+import db from './db'
+
+obey.creator('uuidCreator', uuid.v4)
+
+const salt = new Buffer('somesalt')
+obey.modifier('encrypt', val => argon2.hash(val, salt))
+
+// Add a type to ensure we have a valid, unique email
+obey.type('uniqueEmail', context => {
+    // Read from the 'db' datasource where email is the value
+    return db.read({ email: context.val }).then(res => {
+        // Fail if it already exists
+        if (res.length !== 0) context.fail('Email must be unique')
+        // Note: you'd also want to check that the email is valid!
+    })
+})
+
+const user = obey.model({
+    id: { type: 'uuid', creator: 'uuidCreator' required: true },
+    // Email is now set to type 'uniqueEmail'
+    email: { type: 'uniqueEmail', required: true },
+    password: { type: 'string', modifier: 'encrypt', require: true }
+})
+{% endhighlight %}
+
+Now when the `validate` method is run we do a read on the datasource and check that the email doesn't already exist. Obey doesn't need to be returned a value or `true`, if the `context.fail` method is not called it will simply move forward.
+
+What's best is **we do all of this at the data validation level**. Yes, we're still hitting the datasource (really no way around that), but all of our handling of an error condition is in the same place as the rest of the validation error handling.
+
+The code above is also getting a bit unweildy and is probably better abstracted, but Obey is built for that; all of the modifiers, creators, and types can be added in an abstraction then called wherever you create a model. This makes it easy to share types, modifiers, creators, and the like across multiple models.
+
+But wait, that custom type is asynchronous? Why yes it is. The creators and modifiers can be as well...
+
+## Embracing Asynchronicity
+
+It's simple to look at data modelling and validation as a synchronous thing; running through checks and moving on, but it should have more to it than just that. It limits the power of this integral layer of application development.
+
+With everything else in data i/o being asynchronous it's one area where asynchronous processing is not a _nice to have_, it's a **must have**.
```

Comment:
Expansion desirable. Is most data rlly async, or a mix of async and synchronous?